storagenode/piecestore/endpoint_test.go

```diff
@@ -4,31 +4,84 @@
 package piecestore_test
 
 import (
+	"io"
+	"math/rand"
 	"testing"
+	"time"
 
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
+	"storj.io/storj/internal/memory"
 	"storj.io/storj/internal/testcontext"
 	"storj.io/storj/internal/testplanet"
-	"storj.io/storj/pkg/pb"
+	"storj.io/storj/storagenode/bandwidth"
+	"storj.io/storj/uplink/piecestore"
 )
 
-func TestEndpointConnect(t *testing.T) {
+func TestUploadAndPartialDownload(t *testing.T) {
 	ctx := testcontext.New(t)
 	defer ctx.Cleanup()
 
-	planet, err := testplanet.New(t, 0, 1, 1)
+	planet, err := testplanet.New(t, 1, 6, 1)
 	require.NoError(t, err)
 	defer ctx.Check(planet.Shutdown)
 
 	planet.Start(ctx)
 
-	client, err := planet.Uplinks[0].DialPiecestore(ctx, planet.StorageNodes[0])
+	expectedData := make([]byte, 100*memory.KiB)
+	_, err = rand.Read(expectedData)
 	require.NoError(t, err)
-	defer ctx.Check(client.Close)
 
-	// TODO: at the moment should fail because there's no signature and other things are wrong with it
-	err = client.Delete(ctx, &pb.OrderLimit2{})
-	t.Log(err)
+	err = planet.Uplinks[0].Upload(ctx, planet.Satellites[0], "test/bucket", "test/path", expectedData)
+	assert.NoError(t, err)
+
+	var totalDownload int64
+	for _, tt := range []struct {
+		offset, size int64
+	}{
+		{0, 1510},
+		{1513, 1584},
+		{13581, 4783},
+	} {
+		if piecestore.DefaultConfig.InitialStep < tt.size {
+			t.Fatal("test expects initial step to be larger than size to download")
+		}
+		totalDownload += piecestore.DefaultConfig.InitialStep
+
+		download, err := planet.Uplinks[0].DownloadStream(ctx, planet.Satellites[0], "test/bucket", "test/path")
+		require.NoError(t, err)
+
+		pos, err := download.Seek(tt.offset, io.SeekStart)
+		require.NoError(t, err)
+		assert.Equal(t, pos, tt.offset)
+
+		data := make([]byte, tt.size)
+		n, err := io.ReadFull(download, data)
```

Comment:
does it make sense to make `data` larger than `tt.size` to check if will read only `tt.size`?

Reply:
not really, because the `io.ReadFull` reads the tt.size. Since there's no ranger, there's no way to set the upper limit here.